pkg/control/sidecarcontrol/util.go

```diff
@@ -270,13 +274,80 @@ func GetSidecarTransferEnvs(sidecarContainer *appsv1alpha1.SidecarContainer, pod
-		if !ok {
-			klog.Warningf("there is no env %v in container %v", tEnv.EnvName, tEnv.SourceContainerName)
-			continue
+		envs := sets.NewString()
+		if tEnv.EnvName != "" {
+			envs.Insert(tEnv.EnvName)
+		}
+		for _, e := range tEnv.EnvNames {
+			envs.Insert(e)
+		}
+
+		sourceContainerName := tEnv.SourceContainerName
+		if tEnv.SourceContainerNameFrom != nil && tEnv.SourceContainerNameFrom.FieldRef != nil {
+			containerName, err := ExtractContainerNameFromFieldPath(tEnv.SourceContainerNameFrom.FieldRef, pod)
+			if err != nil {
+				klog.Errorf("unmarshal pod(%s.%s) annotations[%s] failed: %s", pod.Namespace, pod.Name, err.Error())
+				continue
+			}
+			sourceContainerName = containerName
+		}
+		// if sourceContainerName is empty or not found in pod.spec.containers
+		for _, envName := range envs.List() {
+			key := fmt.Sprintf("%v/%v", sourceContainerName, envName)
+			env, ok := envsInPod[key]
+			if !ok {
+				klog.Warningf("there is no env %v in container %v", tEnv.EnvName, tEnv.SourceContainerName)
+				continue
+			}
+			injectedEnvs = append(injectedEnvs, env)
 		}
-		injectedEnvs = append(injectedEnvs, env)
 	}
 	return
 }
+
+func ExtractContainerNameFromFieldPath(fs *corev1.ObjectFieldSelector, pod *corev1.Pod) (string, error) {
+	fieldPath := fs.FieldPath
+	accessor, err := meta.Accessor(pod)
+	if err != nil {
+		return "", err
+	}
+	path, subscript, ok := fieldpath.SplitMaybeSubscriptedPath(fieldPath)
+	if ok {
+		switch path {
+		case "metadata.annotations":
+			if errs := validation.IsQualifiedName(strings.ToLower(subscript)); len(errs) != 0 {
+				return "", fmt.Errorf("invalid key subscript in %s: %s", fieldPath, strings.Join(errs, ";"))
+			}
+			return accessor.GetAnnotations()[subscript], nil
+		case "metadata.labels":
+			if errs := validation.IsQualifiedName(subscript); len(errs) != 0 {
+				return "", fmt.Errorf("invalid key subscript in %s: %s", fieldPath, strings.Join(errs, ";"))
+			}
+			return accessor.GetLabels()[subscript], nil
+		default:
+			return "", fmt.Errorf("fieldPath %q does not support subscript", fieldPath)
+		}
+	}
+	return "", fmt.Errorf("unsupported fieldPath: %v", fieldPath)
+}
+
+// code lifted from https://github.com/kubernetes/kubernetes/blob/master/pkg/apis/core/pods/helpers.go
+// ConvertDownwardAPIFieldLabel converts the specified downward API field label
+// and its value in the pod of the specified version to the internal version,
+// and returns the converted label and value. This function returns an error if
+// the conversion fails.
+func ConvertDownwardAPIFieldLabel(version, label, value string) (string, string, error) {
+	if version != "v1" {
+		return "", "", fmt.Errorf("unsupported pod version: %s", version)
+	}
+	path, _, ok := fieldpath.SplitMaybeSubscriptedPath(label)
+	if ok {
+		switch path {
+		case "metadata.annotations", "metadata.labels":
+			return label, value, nil
+		default:
+			return "", "", fmt.Errorf("field label does not support subscript: %s", label)
```

Comment:
the error message seems a bit confusing, consider replace the  error "field path not supported: %s", path